### ISSUE TITLE
Flag variables that 1. aren't reassigned, and 2. are missing `Final`

### ISSUE THREAD

**Issue by hamirmahal** [astral-sh/ruff#21584]
### Summary

I think it'd be helpful if `ruff` flagged variables that

1. aren't reassigned, and
2. are missing `Final`

**Comment by hirehamir** [comment]
For example,

```python
X: Final = 1
print(X)
```

is fine, but

```python
X = 1
print(X)
```
would result in a warning

**Comment by MichaReiser** [comment]
Can you say more about where this rule would apply? Applying to all variables seems problematic as a variable can be mutated from outside the module. 

I'm also not sure what the value is within a single function. It seems overly pedantic to me with questionable value (reminding me of the JavaScript `let` vs `const`)

I did take a quick look at your PR and the high number of violations strongly suggests that this rule needs further refinement.

**Comment by hirehamir** [comment]
> I'm also not sure what the value is within a single function. It seems overly pedantic to me with questionable value (reminding me of the JavaScript `let` vs `const`)

It prevents accidental reassignments.
